File: causa/src/agents/telemetry.py
"""
telemetry.py — Step 5: cost/token telemetry (task §20).

Two kinds of TelemetryRecord get appended to InvestigationState.telemetry:
  - record_llm_call(): one real Groq API call by an LLM-backed agent
    (Hypothesis/Evidence/Counter-Evidence) -- real model id, real
    input/output tokens (from the provider's own `usage` field), a real
    estimated cost from GROQ_PRICING_PER_MILLION_TOKENS.
  - record_deterministic_call(): one pass through a fully deterministic
    agent (Orchestrator/CausalSelector/ConfidenceJudge, or a tool-only call)
    -- model="deterministic_rule_engine_v1", 0 tokens, 0 cost, because that
    is genuinely true for these agents (see agents/models.py's module
    docstring for the full split rationale).

Pricing is APPROXIMATE and will drift as Groq's published rates change --
treat GROQ_PRICING_PER_MILLION_TOKENS as a best-effort estimate for the
prototype's cost telemetry, not a billing-accurate source of truth. An
unknown model id (or Groq's actual free-tier billing for these specific
keys, which may be $0 regardless of the table below) reports cost as 0.0
rather than guessing.
"""

from __future__ import annotations

from typing import Any

from agents.models import AgentRole, InvestigationState, TelemetryRecord

# USD per 1,000,000 tokens. Source: approximate published Groq rates at the
# time this was written; re-check https://groq.com/pricing before treating
# any number here as billing-accurate.
GROQ_PRICING_PER_MILLION_TOKENS: dict[str, dict[str, float]] = {
    # Verified live-reachable with the user's supplied keys and real
    # tool-calling support (see STEP5_VALIDATION.md §14/§17) -- "openai/gpt-oss-20b"
    # is the actual default (agents.llm_client.DEFAULT_MODEL).
    "openai/gpt-oss-20b": {"input": 0.10, "output": 0.50},
    "openai/gpt-oss-120b": {"input": 0.15, "output": 0.75},
    "compound-beta": {"input": 0.59, "output": 0.79},
    "compound-beta-mini": {"input": 0.10, "output": 0.10},
    # Kept for completeness / possible future availability -- NOT verified
    # reachable with this user's keys (see docs/AGENT_SECURITY.md §5).
    "llama-3.3-70b-versatile": {"input": 0.59, "output": 0.79},
    "llama-3.1-8b-instant": {"input": 0.05, "output": 0.08},
    "llama-3.1-70b-versatile": {"input": 0.59, "output": 0.79},
    "mixtral-8x7b-32768": {"input": 0.24, "output": 0.24},
}
_DEFAULT_PRICE = {"input": 0.0, "output": 0.0}


def estimate_cost(model: str, input_tokens: int, output_tokens: int) -> float:
    price = GROQ_PRICING_PER_MILLION_TOKENS.get(model, _DEFAULT_PRICE)
    return round(input_tokens / 1_000_000 * price["input"] + output_tokens / 1_000_000 * price["output"], 6)


def record_llm_call(state: InvestigationState, agent_role: AgentRole, response: Any,
                     latency_ms: float) -> TelemetryRecord:
    rec = TelemetryRecord(
        agent_role=agent_role.value, model=response.model, input_tokens=response.input_tokens,
        output_tokens=response.output_tokens, total_tokens=response.input_tokens + response.output_tokens,
        estimated_cost=estimate_cost(response.model, response.input_tokens, response.output_tokens),
        tool_calls=0, retrieval_calls=0, agent_latency_ms=round(latency_ms, 3), total_latency_ms=round(latency_ms, 3),
    )
    state.telemetry.append(rec)
    return rec


def record_deterministic_call(state: InvestigationState, agent_role: AgentRole, latency_ms: float,
                               tool_calls: int = 0, retrieval_calls: int = 0) -> TelemetryRecord:
    rec = TelemetryRecord(
        agent_role=agent_role.value, model="deterministic_rule_engine_v1", input_tokens=0, output_tokens=0,
        total_tokens=0, estimated_cost=0.0, tool_calls=tool_calls, retrieval_calls=retrieval_calls,
        agent_latency_ms=round(latency_ms, 3), total_latency_ms=round(latency_ms, 3),
    )
    state.telemetry.append(rec)
    return rec


def aggregate(state: InvestigationState) -> dict:
    total_input = sum(t.input_tokens for t in state.telemetry)
    total_output = sum(t.output_tokens for t in state.telemetry)
    return {
        "total_llm_calls": sum(1 for t in state.telemetry if t.model != "deterministic_rule_engine_v1"),
        "total_deterministic_calls": sum(1 for t in state.telemetry if t.model == "deterministic_rule_engine_v1"),
        "total_input_tokens": total_input, "total_output_tokens": total_output,
        "total_tokens": total_input + total_output,
        "total_estimated_cost": round(sum(t.estimated_cost for t in state.telemetry), 6),
        "total_tool_calls": len(state.audit_trace),
        "total_retrieval_calls": state.budgets.used_retrieval_calls,
        "total_agent_latency_ms": round(sum(t.agent_latency_ms for t in state.telemetry), 3),
        "by_agent_role": {
            role.value: {
                "calls": sum(1 for t in state.telemetry if t.agent_role == role.value),
                "tokens": sum(t.total_tokens for t in state.telemetry if t.agent_role == role.value),
                "cost": round(sum(t.estimated_cost for t in state.telemetry if t.agent_role == role.value), 6),
            }
            for role in AgentRole
        },
    }

File: frontend/src/api/productionApi/telemetry.ts
import { apiFetch } from './client'
import { getCurrentInvestigationId, getRunMeta } from './investigations'

export interface AgentMetric {
  agentRole: string
  calls: number
  tokens: number
  cost: number
}

export interface TelemetryView {
  runMeta: { generatedAt: string; llmProvider: string; llmModel: string }
  role: 'ANALYST' | 'EXECUTIVE'
  totalLatencyMs: number
  totalTokens: number
  totalCost: number
  llmCalls: number
  deterministicCalls: number
  toolCalls: number
  retrievalCalls: number
  cacheHits: number
  cacheMisses: number
  byAgentRole: AgentMetric[]
  telemetryAvailable: boolean
}

interface TelemetryResponse {
  telemetry_available: boolean
  total_agent_latency_ms?: number
  total_tokens?: number
  total_estimated_cost?: number
  total_llm_calls?: number
  total_deterministic_calls?: number
  total_tool_calls?: number
  total_retrieval_calls?: number
  by_agent_role?: Record<string, { calls: number; tokens?: number; cost?: number }>
}

export async function getTelemetryView(role: 'ANALYST' | 'EXECUTIVE'): Promise<TelemetryView> {
  const [id, runMeta] = await Promise.all([getCurrentInvestigationId(role), getRunMeta()])
  const r = await apiFetch<TelemetryResponse>(`/api/investigations/${id}/telemetry?requester_role=${role}`)
  const byAgentRole: AgentMetric[] = Object.entries(r.by_agent_role ?? {}).map(([agentRole, v]) => ({
    agentRole, calls: v.calls, tokens: v.tokens ?? 0, cost: v.cost ?? 0,
  }))
  return {
    runMeta, role, telemetryAvailable: r.telemetry_available,
    totalLatencyMs: r.total_agent_latency_ms ?? 0, totalTokens: r.total_tokens ?? 0, totalCost: r.total_estimated_cost ?? 0,
    llmCalls: r.total_llm_calls ?? 0, deterministicCalls: r.total_deterministic_calls ?? 0,
    toolCalls: r.total_tool_calls ?? 0, retrievalCalls: r.total_retrieval_calls ?? 0,
    cacheHits: 0, cacheMisses: 0, byAgentRole,
  }
}
